editor/sidebar/post-taxonomies/index.js

```diff
@@ -1,24 +1,74 @@
 /**
+ * External Dependencies
+ */
+import { connect } from 'react-redux';
+
+/**
  * WordPress dependencies
  */
 import { __ } from '@wordpress/i18n';
 import { PanelBody } from '@wordpress/components';
+import { Component } from '@wordpress/element';
 
 /**
  * Internal dependencies
  */
 import './style.scss';
-import TagsSelector from './tags-selector';
-import CategoriesSelector from './categories-selector';
-
-function PostTaxonomies() {
-	return (
-		<PanelBody title={ __( 'Categories & Tags' ) } initialOpen={ false }>
-			<CategoriesSelector />
-			<TagsSelector />
-		</PanelBody>
-	);
+import HierarchicalTaxonomy from './hierarchical-taxonomy';
+import FlatTaxonomy from './flat-taxonomy';
+import { getCurrentPostType } from '../../selectors';
+
+class PostTaxonomies extends Component {
+	constructor() {
+		super( ...arguments );
+		this.state = {
+			taxonomies: [],
+		};
+	}
+
+	componentDidMount() {
+		this.fetchTaxonomies = new wp.api.collections.Taxonomies()
+			.fetch()
+			.done( ( taxonomies ) => {
+				this.setState( { taxonomies: Object.values( taxonomies ) } );
+			} );
+	}
+
+	componentWillUnmout() {
+		this.fetchTaxonomies.abort();
+	}
+
+	render() {
+		const availableTaxonomies = this.state.taxonomies
+			.filter( ( taxonomy ) => taxonomy.types.indexOf( this.props.postType ) !== -1 );
+
+		if ( ! availableTaxonomies.length ) {
+			return null;
+		}
+
+		return (
+			<PanelBody title={ __( 'Categories & Tags' ) } initialOpen={ false }>
+				{ availableTaxonomies.map( ( taxonomy ) => {
+					const TaxonomyComponent = taxonomy.hierarchical ? HierarchicalTaxonomy : FlatTaxonomy;
+					return (
+						<TaxonomyComponent
+							key={ taxonomy.name }
```

Comment:
Since `name` is human-readable label and not necessarily unique, should we use `taxonomy.slug` instead?